src/arcsf/data/tofu.py

```diff
@@ -0,0 +1,193 @@
+from sklearn.model_selection import train_test_split
+
+
+def _flatten_list_of_lists(list: list[list]) -> list:
+    return sum(list, [])
+
+
+# returns the question index in the dataset, given an author number and question
+# number (within the author) (inner function)
+def _get_forget_index(
+    author: int,
+    question: int,
+    q_per_author: int,
+) -> int:
+    return question + q_per_author * author
+
+
+# As above returns the question indices in the dataset, given an
+# author number and question numbers (within the author) (inner function)
+def _get_forget_indices(
+    authors: int,
+    questions: int | list[int],
+    q_per_author: int,
+) -> int | list[int]:
+    if isinstance(questions, int):
+        return _get_forget_index(authors, questions, q_per_author)
+    if isinstance(questions, list):
+        return [
+            _get_forget_index(authors, question, q_per_author) for question in questions
+        ]
+
+
+# As above returns the question indices in the dataset, given
+# author numbers and question numbers (across dataset)
+def get_forget_indices(
+    authors: int | list[int],
+    questions: int | list[int],
+    q_per_author: int,
+) -> list[int]:
+    if isinstance(authors, int):
+        return _get_forget_indices(authors, questions, q_per_author)
+    if isinstance(authors, list):
+        out = [
+            _get_forget_indices(author, questions, q_per_author) for author in authors
+        ]
+        if isinstance(out[0], list):
+            return _flatten_list_of_lists(out)
+        return out
+
+
+def load_tofu(
+    granularity: str,
+    stratified: bool,
+    forget_random: bool,
+    forgotten_author_fraction: float,
+    forgotten_fact_fraction: float,
+    random_seed: int,
+    debug=False,
+) -> tuple[Dataset, Dataset, dict] | tuple[Dataset, Dataset]:
+    """
+    Loads TOFU dataset given different flags for retain--forget split.
+    Args:
+        - granularity: level at which forgetting takes place (author vs question)
+        - stratified: if forgetting questions restrain to specific authors?
+        - forget_random: is forgetting happening randomly within constraints?
+        - forgotten_author_fraction: fraction of authors from which to forget questions
+        - forgotten_fact_fraction: fraction of questions to randomly forget
+            - if stratified == True represents fraction of Qs in author
+            - if stratified == False represents fraction of total Qs
+        - random_seed: seed for reproducibility
+        - debug: returns dictionary containing meta_data if being used during testing
+    Returns:
+        - Two datasets with forget and retain sets
+            as well as a debugging dictionary (optional)
+    """
+
+    all_data = load_dataset("locuslab/TOFU", "full")["train"]
+
+    num_authors = 200  # hard coding author count for now
+    q_per_author = 20  # hard coding author question count for now
+
+    q_indices = [q for q in range(num_authors * q_per_author)]
```

Comment:
not sure but maybe should be a per-author index, e.g. author 0 question 0, ... author 0 question 19, author 1 question 0, author 19 question 19, ... etc.? Or have both possibly (an overall row index and a question index for which number question it is for that author)

Reply:
I think I will keep it like this for now